File: crates/flighthq-spritesheet/src/spritesheet_timeline_source.rs
//! `create_spritesheet_timeline_source` — exposes a spritesheet animation as a
//! `TimelineSource` so a MovieClip timeline can play it.
//!
//! This is the spritesheet side of the timeline frame-source contract:
//! `flighthq-timeline` consumes a `TimelineSource`, this produces one, and
//! neither crate depends on the other beyond the shared `TimelineSource` type
//! in `flighthq-types`.
//!
//! In the TS SDK each frame swaps the displayed atlas region and offset on a
//! `Bitmap` child the source lazily creates on the target display object. The
//! Rust `TimelineSource.construct_frame` callback receives only the opaque
//! target node id (`u64`) and the 1-based frame number — it has no access to a
//! display-object arena — so this port resolves the per-frame placement (the
//! atlas region plus the offset, with the animation origin subtracted) and
//! forwards it to a caller-supplied `apply` callback. The caller wires that
//! placement onto a bitmap in their own display graph, which keeps the
//! spritesheet crate decoupled from `flighthq-displayobject`.

use flighthq_types::{Spritesheet, SpritesheetAnimation, TextureAtlasRegion, TimelineSource};

/// The resolved display state for a single spritesheet timeline frame: which
/// atlas region to show and where to position it. Mirrors the per-frame
/// `bitmap.data.sourceRectangle` / `bitmap.x` / `bitmap.y` writes the TS
/// `constructFrame` performs.
#[derive(Clone, Debug)]
pub struct SpritesheetFramePlacement {
    pub region: TextureAtlasRegion,
    pub x: f32,
    pub y: f32,
}

/// Builds a [`TimelineSource`] that plays `animation` from `spritesheet`.
///
/// `total_frames` is the animation's frame count and `frame_rate` is
/// `1000 / frame_duration`, matching the TS source exactly. On each frame the
/// source resolves the spritesheet frame (`spritesheet.frames[animation.frames
/// [frame - 1]]`), looks up its atlas region, subtracts the animation origin
/// from the frame offset, and calls `apply(target_id, placement)`.
///
/// `apply` is a no-op when the spritesheet has no atlas, when the animation
/// frame index is out of range, or when the resolved frame has no matching
/// atlas region — mirroring the early returns in the TS `constructFrame`.
pub fn create_spritesheet_timeline_source(
    spritesheet: Spritesheet,
    animation: SpritesheetAnimation,
    apply: Box<dyn Fn(u64, SpritesheetFramePlacement) + Send + Sync>,
) -> TimelineSource {
    let total_frames = animation.frames.len() as u32;
    let frame_rate = Some(1000.0 / animation.frame_duration);
    TimelineSource {
        total_frames,
        frame_rate,
        labels: Vec::new(),
        construct_frame: Box::new(move |target_id, frame| {
            let Some(atlas) = spritesheet.atlas.as_ref() else {
                return;
            };
            let Some(frame_id) = animation.frames.get((frame - 1) as usize).copied() else {
                return;
            };
            let Some(sheet_frame) = spritesheet.frames.get(frame_id as usize) else {
                return;
            };
            let Some(region) = atlas.regions.iter().find(|r| r.id == sheet_frame.id) else {
                return;
            };
            apply(
                target_id,
                SpritesheetFramePlacement {
                    region: region.clone(),
                    x: sheet_frame.offset_x - animation.origin_x,
                    y: sheet_frame.offset_y - animation.origin_y,
                },
            );
        }),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::spritesheet::{CreateSpritesheetOptions, create_spritesheet};
    use crate::spritesheet_animation::{
        CreateSpritesheetAnimationOptions, create_spritesheet_animation,
    };
    use crate::spritesheet_frame::{CreateSpritesheetFrameOptions, create_spritesheet_frame};
    use flighthq_textureatlas::{add_texture_atlas_region, create_texture_atlas};
    use std::sync::{Arc, Mutex};

    fn make_sheet(frame_count: u32) -> Spritesheet {
        let mut atlas = create_texture_atlas(None, Vec::new());
        let mut frames = Vec::new();
        for i in 0..frame_count {
            add_texture_atlas_region(&mut atlas, (i * 32) as f32, 0.0, 32.0, 32.0, None, None);
            frames.push(create_spritesheet_frame(CreateSpritesheetFrameOptions {
                id: Some(i),
                ..Default::default()
            }));
        }
        create_spritesheet(CreateSpritesheetOptions {
            atlas: Some(atlas),
            frames: Some(frames),
            ..Default::default()
        })
    }

    #[test]
    fn create_spritesheet_timeline_source_reports_total_frames_and_frame_rate() {
        let sheet = make_sheet(3);
        let anim = create_spritesheet_animation(CreateSpritesheetAnimationOptions {
            frame_duration: Some(200.0),
            frames: Some(vec![0, 1, 2]),
            ..Default::default()
        });
        let source = create_spritesheet_timeline_source(sheet, anim, Box::new(|_, _| {}));
        assert_eq!(source.total_frames, 3);
        assert!((source.frame_rate.unwrap() - 1000.0 / 200.0).abs() < 1e-4);
    }

    #[test]
    fn create_spritesheet_timeline_source_resolves_the_frame_region_and_offset() {
        let sheet = make_sheet(2);
        let anim = create_spritesheet_animation(CreateSpritesheetAnimationOptions {
            frame_duration: Some(100.0),
            frames: Some(vec![0, 1]),
            ..Default::default()
        });
        let placements = Arc::new(Mutex::new(Vec::<(u64, SpritesheetFramePlacement)>::new()));
        let sink = placements.clone();
        let source = create_spritesheet_timeline_source(
            sheet,
            anim,
            Box::new(move |id, placement| sink.lock().unwrap().push((id, placement))),
        );
        (source.construct_frame)(7, 1);
        (source.construct_frame)(7, 2);
        let got = placements.lock().unwrap();
        assert_eq!(got.len(), 2);
        assert_eq!(got[0].0, 7);
        assert_eq!(got[0].1.region.x, 0.0);
        assert_eq!(got[1].1.region.x, 32.0);
    }

    #[test]
    fn create_spritesheet_timeline_source_subtracts_the_animation_origin() {
        let mut sheet = make_sheet(1);
        sheet.frames[0].offset_x = 10.0;
        sheet.frames[0].offset_y = 20.0;
        let anim = create_spritesheet_animation(CreateSpritesheetAnimationOptions {
            frame_duration: Some(100.0),
            frames: Some(vec![0]),
            origin_x: Some(4.0),
            origin_y: Some(6.0),
            ..Default::default()
        });
        let placements = Arc::new(Mutex::new(Vec::<SpritesheetFramePlacement>::new()));
        let sink = placements.clone();
        let source = create_spritesheet_timeline_source(
            sheet,
            anim,
            Box::new(move |_, placement| sink.lock().unwrap().push(placement)),
        );
        (source.construct_frame)(0, 1);
        let got = placements.lock().unwrap();
        assert_eq!(got[0].x, 6.0);
        assert_eq!(got[0].y, 14.0);
    }

    #[test]
    fn create_spritesheet_timeline_source_does_not_apply_when_the_spritesheet_has_no_atlas() {
        let sheet = create_spritesheet(CreateSpritesheetOptions::default());
        let anim = create_spritesheet_animation(CreateSpritesheetAnimationOptions {
            frame_duration: Some(100.0),
            frames: Some(vec![0]),
            ..Default::default()
        });
        let called = Arc::new(Mutex::new(false));
        let sink = called.clone();
        let source = create_spritesheet_timeline_source(
            sheet,
            anim,
            Box::new(move |_, _| *sink.lock().unwrap() = true),
        );
        (source.construct_frame)(0, 1);
        assert!(!*called.lock().unwrap());
    }
}

File: packages/movieclip/src/spritesheetTimelineSource.ts
import { createEntity } from '@flighthq/entity/contract';
import { addNodeChild, invalidateNodeLocalTransform } from '@flighthq/node/contract';
import { createSprite } from '@flighthq/scene2d/contract';
import { getTextureAtlasRegionTexture } from '@flighthq/textureatlas/contract';
import type {
  Entity,
  EntityWithoutRuntime,
  Node2D,
  Sprite,
  Spritesheet,
  SpritesheetAnimation,
  SpritesheetTimelineSourceExplanation,
  SpritesheetTimelineSourceGuard,
  TimelineSource,
} from '@flighthq/types/contract';

// Exposes a spritesheet animation as a TimelineSource so a MovieClip can play it (the spritesheet side of
// the timeline frame-source contract — `@flighthq/timeline` consumes `TimelineSource`, this produces one,
// and neither depends on the other beyond the shared interface in `@flighthq/types`). Each frame swaps
// the displayed atlas region and offset on a bitmap the source lazily creates as a child of the target
// the first time it constructs onto that target; tracking the bitmap per target keeps the source
// shareable across many MovieClips.
//
// Bind it with `setMovieClipSource(clip, createSpritesheetTimelineSource(sheet, anim))`, then `playMovieClip`.
export function createSpritesheetTimelineSource(
  spritesheet: Readonly<Spritesheet>,
  animation: Readonly<SpritesheetAnimation>,
): TimelineSource & Entity {
  const bitmaps = new WeakMap<Node2D, Sprite>();
  const frames = materializeSpritesheetTimelineFrames(animation);
  if (_spritesheetTimelineSourceGuard !== null) {
    _spritesheetTimelineSourceGuard(animation, explainSpritesheetTimelineSource(animation));
  }
  return createEntity<EntityWithoutRuntime<TimelineSource>>({
    totalFrames: frames.length,
    labels: [],
    // A spritesheet animation is pure frame content; the format carries nothing to cue.
    cues: [],
    frameRate: 1000 / animation.frameDuration,
    constructFrame(target: Node2D, frame: number): void {
      const atlas = spritesheet.atlas;
      if (atlas === null) return;

      let bitmap = bitmaps.get(target);
      if (bitmap === undefined) {
        bitmap = createSprite();
        addNodeChild(target, bitmap);
        bitmaps.set(target, bitmap);
      }

      const sheetFrame = spritesheet.frames[frames[frame - 1]];
      if (sheetFrame === undefined) return;
      bitmap.data.texture = getTextureAtlasRegionTexture(atlas, sheetFrame.id);
      bitmap.x = sheetFrame.offsetX - animation.originX;
      bitmap.y = sheetFrame.offsetY - animation.originY;
      invalidateNodeLocalTransform(bitmap);
    },
  });
}

// Reports the authored playback fields the TimelineSource vocabulary cannot carry. Direction is exact:
// createSpritesheetTimelineSource expands reverse and ping-pong directions into the source frame list.
export function explainSpritesheetTimelineSource(
  animation: Readonly<SpritesheetAnimation>,
): SpritesheetTimelineSourceExplanation {
  return {
    directionMaterialized: true,
    unsupportedFields:
      animation.frameDurations === null ? REPEAT_COUNT_UNSUPPORTED : FRAME_DURATIONS_AND_REPEAT_COUNT_UNSUPPORTED,
  };
}

// Diagnostics seam used by enableMovieClipGuards. Null restores the zero-message production path.
export function setSpritesheetTimelineSourceGuard(guard: SpritesheetTimelineSourceGuard | null): void {
  _spritesheetTimelineSourceGuard = guard;
}

// Converts the animation's direction into an ordinary forward TimelineSource frame sequence. A loop over
// [0,1,2,1] has the same boundary as spritesheet ping-pong: its next frame is 0, with neither endpoint
// duplicated. Reverse ping-pong follows the same rule from the other endpoint.
function materializeSpritesheetTimelineFrames(animation: Readonly<SpritesheetAnimation>): readonly number[] {
  const frames = animation.frames;
  if (frames.length < 2 || animation.direction === 'forward') return frames;
  if (animation.direction === 'reverse') return [...frames].reverse();

  const out: number[] = [];
  if (animation.direction === 'pingpong') {
    for (let index = 0; index < frames.length; index++) out.push(frames[index]);
    for (let index = frames.length - 2; index > 0; index--) out.push(frames[index]);
    return out;
  }

  for (let index = frames.length - 1; index >= 0; index--) out.push(frames[index]);
  for (let index = 1; index < frames.length - 1; index++) out.push(frames[index]);
  return out;
}

const FRAME_DURATIONS_AND_REPEAT_COUNT_UNSUPPORTED = ['frameDurations', 'repeatCount'] as const;
const REPEAT_COUNT_UNSUPPORTED = ['repeatCount'] as const;
let _spritesheetTimelineSourceGuard: SpritesheetTimelineSourceGuard | null = null;
